[Friday] Navigating many to one relations blog (#102)
xataio/mdx-blog: navigating-many-to-one.mdx

---
title: 'Managing many-to-one and one-to-many database relations'

> description: 'Addressing the n+1 problem and navigating one-to-many relationships with Xata's new approach.'

Reviewer: The `'` is breaking the string, that's why the preview is not building.

```suggestion
description: "Addressing the n+1 problem and navigating one-to-many relationships with Xata's new approach."
```

image:
  src: 'https://raw.githubusercontent.com/xataio/mdx-blog/main/images/csv-import-export-cover.jpg'
  alt: 'Many-to-one relationship'
author: Alejandro Martínez Vieites
date: 09-01-2023
published: true

> slug: navigating-many-to-one

Reviewer 2: When we have the image, let's drop in an `ogImage` attribute here to use Elizabets design. Looks like we have one here: https://github.com/xataio/mdx-blog/issues/23#issuecomment-1701330014

---

As our [launch week](/blog/launch-week-august-2023) comes to a close, we're looking at APIs and their intricate relationships. Today, we're introducing a new syntax that simplifies navigating these relationships. While Xata already provides a great way to retrieve related data using [link columns](/docs/concepts/data-model#link) for the _many-to-one_ relationship, we're addressing the challenge of navigating the reverse [relationship](https://xata.io/docs/concepts/data-model#links-and-relations) (_one-to-many_). This new approach improves efficiency and helps avoid the _n + 1 problem_.

Let's consider a basic schema. Within the schema, there's a _posts_ table containing certain information and establishing a connection to the corresponding author. This author-related data belongs in the _users_ table.

> ![](image/many-to-one-relationships.png)

Reviewer: ```suggestion
![Xata schema showing a many-to-one relationship](image/many-to-one-relationships.png)
```

> When querying the _posts_ table we are able to retrieve the title, likes, and the author's name (and any other fields present in the _users_ table):

Reviewer: If we don't plan to show SDK calls for these, we can introduce this section with something like:

> To explain the N+1 problem, we're going to show what API requests are sent of the network. This is what the Xata SDKs are doing behind the scenes.

Reviewer 2: If we can show the SDK calls similar to other blogs, I really think we should. @SferaDev @philkra I'm pretty sure we have SDK support, how hard would it be to match some of these examples?

Reviewer: In this case, because we're showing the responses as well and because we're explaining the problem of round-trips, showing only the API calls might be better.

However, how to call the reverse links via the SDK somewhere in the blog would be really good.

Reviewer 3: TS & Python SDK snippets are in

        }
    ]
}
```

And then, for each user that has been returned, make a separate request in the _posts_ table to retrieve the posts from each author, similar to:

```json
// Request
POST /db/blogs:main/tables/posts/query
{
  "columns": ["*"],
  "filter": {
      "author": "rec_cie05krjtojbm41fv2q0"
  }
}

// Response (Xata fields have been omitted for clarity)
{
    "records": [
        {
            "author": {
                "id": "rec_cie05krjtojbm41fv2q0"
            },
            "id": "rec_cie05srjtojbm41fv2t0",
            "likes": 13,
            "title": "Postgres schema changes are still a PITA"
        },
        {
            "author": {
                "id": "rec_cie05krjtojbm41fv2q0"
            },
            "id": "rec_cie05v3jtojbm41fv2tg",
            "likes": 15,
            "title": "On the performance impact of REPLICA IDENTITY FULL in Postgres"
        },
        {
            "author": {
                "id": "rec_cie05krjtojbm41fv2q0"
            },
            "id": "rec_cii1kajjtoj0t82ae3ag",
            "likes": 6,
            "title": "Semantic or keyword search for finding ChatGPT context. Who searched it better?"
        }
    ]
}
```

> We are performing "1" (get the authors) + "n" (get the authors's posts) requests. This is inconvenient and also quite inefficient, since the number of requests depends on the number of user records.

Reviewer: Capital N for consistency?

```suggestion
We are performing "1" (get the authors) + "N" (get the authors's posts) requests. This is inconvenient and also quite inefficient, since the number of requests depends on the number of user records.
```

There are workarounds depending on the data set and the amount of code we are willing to write, like getting _all_ the authors and _all_ the posts, or using the `IN` operator and then processing the response in the app. But it would be better if our API would help us getting exactly what we want.

# Navigate links backwards

> Since the relationship is already defined by the `link` column, it’s a matter of providing a good API experience to get this data more easily. We have introduced [column expressions](/docs/concepts/data-model#links-and-relations) in the query endpoint so the `columns` field is no longer restricted to a list of column names. Now we are able to provide a JSON object to better define the projection we want.

Reviewer 2: Just noticed the docs only has Python and REST examples. If there's TS SDK support, we should update too.

            }
        },
        {
            "id": "rec_cie05ljjtojbm41fv2qg",
            "name": "Joan Edwards",
            "posts": {
                "records": [
                    {
                        "id": "rec_cie060jjtojbm41fv2u0",
                        "title": "Introducing the new Xata regions: Sydney & Frankfurt"
                    }
                ]
            }
        },
        {
            "id": "rec_cii1ki3jtoj0t82ae3b0",
            "name": "Alex Francoeur",
            "posts": {
                "records": [
                    {
                        "id": "rec_cii1kjjjtoj0t82ae3bg",
                        "title": "The next era of databases are serverless, adaptive, and collaborative"
                    }
                ]
            }
        }
    ]
}
```

For the moment, sorting options and classical pagination (limit and offset) are available for the reverse link querying. In the future, we will add support for filtering and cursor pagination. Similarly, we plan to add support for following multiple reverse links transitively. Nevertheless, we think that the majority of use cases are solved by the current functionality already.

> 

Reviewer 4: I didn't want to go into much detail (and also a bit late for that) but take a look at the paragraph I just added :D

> 

Reviewer 2: This looks great @amvieites ! Thank you for adding it.

# Conclusion

The new Xata querying syntax makes it possible to traverse the N:1 relationships backwards and therefore solves the N + 1 problem in a single round-trip.

> If you have feedback or questions, you can reach out to us on [Discord](https://xata.io/discord) or [X | Twitter](https://twitter.com/xata).

Reviewer 2: Suggestion: X / Twitter

---

> title: 'Managing many-to-one and one-to-many database relations'

Reviewer 2: I don't love these, but brainstorming

"Reducing query round trips with reverse links"
"Reducing query round trips with 1:N and N:1 database relations"
"Reduce query round trips with improved one-to-many relations"

> title: 'Managing many-to-one and one-to-many database relations'

Reviewer: Perhaps something like: "The N+1 problem and walking reverse links backwards"

      "postsauthor": {
        "records": [
          {
            "id": "rec_cie060jjtojbm41fv2u0",
            "title": "Introducing the new Xata regions: Sydney & Frankfurt"
          }
        ]
      }
    },
    {
      "id": "rec_cii1ki3jtoj0t82ae3b0",
      "name": "Alex Francoeur",
      "postsauthor": {
        "records": [
          {
            "id": "rec_cii1kpbjtoj0t82ae3c0",
            "title": "End-to-end preview deployment workflows with Xata and Vercel"
          },
          {
            "id": "rec_cii1kjjjtoj0t82ae3bg",
            "title": "The next era of databases are serverless, adaptive, and collaborative"
          },
          {
            "id": "rec_cii1l1rjtoj0t82ae3cg",
            "title": "Modern database workflows with GitHub, Vercel, Netlify, and Xata"
          }
        ]
      }
    }
  ]
}
```

Using this syntax we are able to avoid the N + 1 problem - getting all the data we want using a single request and, therefore, avoiding any unnecessary roundtrips and repetitive code.

More fields are available to configure the projection further: `as`, `sort`, `limit`, `offset`. The same request can be modified like the following:

> <TabbedCode tabs={['TypeScript', 'JSON']}>

Reviewer 5: ```suggestion
<TabbedCode tabs={['TypeScript']}>
```

image:
  src: 'https://raw.githubusercontent.com/xataio/mdx-blog/main/images/many-to-one-relation.jpg'
  alt: 'Many to one relationship'
author: Alejandro Martínez Vieites
date: 09-01-2023
published: true
slug: navigating-many-to-one
ogImage: https://raw.githubusercontent.com/xataio/mdx-blog/main/images/many-to-one-relation-cover-og.jpg
---

As our [launch week](/blog/launch-week-august-2023) comes to a close, we're looking at APIs and their intricate relationships. Today, we're introducing a new syntax that simplifies navigating these relationships. While Xata already provides a great way to retrieve related data using [link columns](/docs/concepts/data-model#link) for the _many-to-one_ relationship, we're addressing the challenge of navigating the reverse [relationship](/docs/concepts/data-model#links-and-relations) (_one-to-many_). This new approach improves efficiency and helps avoid the _N + 1 problem_.

Let's consider a basic schema. Within the schema, there's a _Posts_ table containing certain information and establishing a connection to the corresponding author. This author-related data belongs in the _Users_ table.

> ![Xata schema showing a many-to-one relationship](images/many-to-one-relationships.png)

Reviewer 2: +1, was going to suggest
